### PR TITLE
Splitting apart db and role creation

### LINKED ISSUE #3: Break database and user role creation apart
Currently you can't have users for the same database in more than one namespace. The role will be created in the second namespace but it won't get any privileges. This can be fixed by simply not returning from `CreateDB` on duplicate database error but I think there are more perks to splitting them apart.

So here's my idea:
```
apiVersion: db.movetokube.com/v1alpha1
kind: Postgres
metadata:
  name: my-db
  namespace: app
spec:
  database: test-db
```

This obviously creates the db `test-db` in postgres and a group role `test-db-group` that becomes the owner of the database.

```
CREATE ROLE test-db-group;
CREATE DATABASE test-db OWNER test-db-group;
```

```
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: test-db-user
  database: test-db
  secretName: test-db-postgres-user
```

This will create a user role `test-db-user` in postgres and then grant `test-db-group` to `test-db-user`.

```
CREATE ROLE test-db-user WITH LOGIN PASSWORD '<random>';
GRANT test-db-group TO test-db-user;
```

The `PostgresUser` CRD could even be extended to give us more powerful options.

```
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: test-db-user
  databases:
  - name: test-db
    privileges:
    - OWNER
  - name: another-db
    privileges:
    - SELECT
    - INSERT
  secretName: test-db-postgres-user
```

```
CREATE ROLE test-db-user WITH LOGIN PASSWORD '<random>';
GRANT test-db-group TO test-db-user;
GRANT SELECT ON DATABASE another-db TO test-db-user;
GRANT INSERT ON DATABASE another-db TO test-db-user;
```

### PR DESCRIPTION
**Pull request by arnarg**
Closes #3 

This splits apart database and user role creation into two CRDs. The issue didn't get any comments so I'm hoping this will at least start a discussion.

```
apiVersion: db.movetokube.com/v1alpha1
kind: Postgres
metadata:
  name: my-db
  namespace: app
spec:
  database: test-db
```

This creates a database called `test-db` and that is owned by role `test-db-group`. If the db and role already exist (the same `Postgres` CR is in another namespace) it's a no-op on the database.

```
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: username
  database: my-db
  secretName: my-credential-secret
```

This creates user role `username-<random-6-chars>` with role `test-db-group` granted to it. There has to be a `Postgres` CR named `my-db` in the same namespace.

If the `PostgresUser` is deleted the user role is simply deleted.

If the `Postgres` is deleted it checks if there is another `Postgres` that is using the same `test-db-group` role (so the same database can be used in multiple namespaces), if not the database ownership is changed to postgres-operator's user and `test-db-group` dropped.

### CODE REVIEW

**Comment by hitman99** [comment]
Hey, I fixed the package names after moving this project to organisation. Could merge these changes to your fork? In the meanwhile, I'll look at this PR in depth

**Comment by arnarg** [comment]
Merged the changes to my fork. Thanks!

**Comment by hitman99** [comment]
I've reviewed your PR and ran some tests. While it works ok if used from scratch, it does not take into consideration that this operator was already deployed and old CRs are present. After deploying an updated operator and deleting a previously created CR, the secret was not deleted, but the role was. This could potentially lead to an application outage

File: pkg/controller/postgres/postgres_controller.go [hunk 1]
```diff
@@ -247,3 +182,27 @@ func (r *ReconcilePostgres) finish(cr *dbv1alpha1.Postgres) (reconcile.Result, e
 	}
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcilePostgres) shouldDropRole(cr *dbv1alpha1.Postgres) bool {
+	// Get a list of all Postgres
+	dbs := dbv1alpha1.PostgresList{}
+	err := r.client.List(context.TODO(), &client.ListOptions{}, &dbs)
+	if err != nil {
+		log.Info(fmt.Sprintf("%v", err))
```
Comment (hitman99) [inline pkg/controller/postgres/postgres_controller.go]: this should reuse reqLogger passed as parameter

File: pkg/postgres/postgres.go [hunk 2]
```diff
@@ -26,51 +32,98 @@ type pg struct {
...
+	_, err := c.db.Exec(fmt.Sprintf(GRANT_ROLE, role, grantee))
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-func (c *pg) DropRole(role, database string, logger logr.Logger) error {
+func (c *pg) AlterDefaultLoginRole(role, setRole string) error {
+	// On AWS RDS the posgres user isn't really superuser so he doesn't have permissions
```
Comment (hitman99) [inline pkg/postgres/postgres.go]: Typo in `postgres`

File: pkg/postgres/postgres.go [hunk 3]
```diff
@@ -26,51 +32,98 @@ type pg struct {
...
+func (c *pg) RevokeRole(role, revoked string) error {
+	_, err := c.db.Exec(fmt.Sprintf(REVOKE_ROLE, role, revoked))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *pg) DropRole(role, newOwner, database string, logger logr.Logger) error {
+	// On AWS RDS the posgres user isn't really superuser so he doesn't have permissions
```
Comment (hitman99) [inline pkg/postgres/postgres.go]: typo in `postgres`

**Comment by hitman99** [comment]
Sorry for picking on the really small things, but after these are fixed, I'll merge this PR and release a new version to operatorhub

**Comment by arnarg** [comment]
No problem, typos make the code look unprofessional ;)
I'm glad you like the proposed changes, should I update the README to reflect them as well?

**Comment by hitman99** [comment]
> I'm glad you like the proposed changes, should I update the README to reflect them as well?
Yes, please. It will be easier for me to update the operator hub then

**Comment by hitman99** [comment]
It's merged, it will take some time for changes to appear in operatorhub, but you could "unofficially" use the operator, as the docker image will be available at `movetokube/postgres-operator:0.3` once the pipeline finishes

**Comment by arnarg** [comment]
Thanks!

**Pull request merged**